tests/blackbox.py

```diff
@@ -440,3 +477,419 @@ def dryrun_one(
+            would differ between model_router and router during the simulation
+        * algod (optional) - if missing, just get one
+
+    Artifacts from Step 1 are stored in self.results: _SimConfig
+
+    2. Simulation (simulate_and_assert method): - using self.results artifacts Step 1, also takes params:
+        * approval_arg_strat_type: Type[ABICallStrategy]
+            - strategy type to use for approval program's arg generation
+        * clear_arg_strat_type_or_inputs: Type[ABICallStrategy] | Iterable[Sequence[PyTypes]] | None
+            - strategy type to use for clear program's arg generation
+        * approval_abi_args_mod: ABIArgsMod (default None)
+            - used to specify any arg mutation
+        # TODO: currently there aren't any clear_abi_args_mod, but we might need these for testing non-trivial clear programs
+        * version: int - for compiling self.router
+        * method_configs: ABICallConfigs - these drive all the test cases
+        * assemble_constants: bool (optional) - for compiling self.router
+        * optimize: OptimizeOptions (optional) - for compiling self.router
+        * num_dryruns: int (default 1)
+            - the number of input runs to generate per method X config combination
+        * txn_params: TxParams (optional)
+            - other TxParams to append in addition to the (is_app_create, OnComplete) information
+        * model_version: int - for compiling self.model_router
+        * model_assemble_constants: bool (optional) - for compiling self.model_router
+        * model_optimize: OptimizeOptions (optional) - for compiling self.model_router
+        * msg: string (optional) - message to report when an assertion is violated
+        * omit_approval_call: bool (default False) - allow purely testing the clear program
+        * omit_clear_call: bool (default False) - allow purely testing the approval program
+        NOTE: one of omit_approval_call or omit_clear_call must remain False
+        * executor_validation (default True) - when False, skip the DryRunExecutor's validation
+        * skip_validation (default False) - when False, skip the Router's validation
+    """
+
+    def __init__(
+        self,
+        router: Router,
+        predicates: CallPredicates,
+        *,
+        model_router: Router | None = None,
+        algod: v2client.algod.AlgodClient | None = None,
+    ):
+        self.router: Router = self._validate_router(router)
+
+        self.predicates: CallPredicates = self._validate_predicates(predicates)
+
+        self.model_router: Router | None = None
+        if model_router:
+            self.model_router = self._validate_router(model_router, kind="Model")
+
+        self.algod: v2client.algod.AlgodClient = algod or algod_with_assertion()
+
+        self.auto_path_results: Dict[
+            str | None, dict[tuple[bool, OnComplete], SimulationResults]
+        ] = _AutoPathDict()
```

Comment:
nit: I am a little bit skeptical about the inherited `defaultdict` here, i.e., `_AutoPathDict`:
- for default value here should be of `dict[tuple[bool, OnComplete], SimulationResults]`
- but it seems the default value is `_AutoPathDict()` by definition of `_AutoPathDict`?

(Also use `dict` not `Dict`)

Reply:
I think I should be able to remove this contraption without too much pain. At the time I introduced it, I didn't know yet how far down I needed to dig for adding various result information. Now that the test construction is complete, I know exactly what I'm inserting. Will keep you posted.

Reply:
ty! I keep an eye on it.

Reply:
see [2e8e06f](https://github.com/algorand/pyteal/pull/634/commits/2e8e06f1b8b79b4bf14f4e1199895b5205198e04)

Reply:
the change looks good, resolving.